### DISCUSSION TITLE
Increase maximum brush size of tools.

### DISCUSSION THREAD

**Discussion by ocalhoun6** [OpenRCT2/OpenRCT2#17072] [Feature requests]
Increase maximum tool brush size of:

- Scenery scatter
- Water raise/lower
- Land raise/lower/paint
- Bulldozer tool

Currently, the maximum size for any of these tools is 64x64. That was fine (though maybe not perfect) for a 255x255 park -- it would still take only a handful of clicks to cover a maximum size park.

But now that the maximum park size is 999x999, it could be very tedious to cover the whole park with any of these tools. About 244 clicks, in the unlikely event that my math is correct. The bulldozer, land paint, and maybe scenery scatter tools can be clicked and dragged, but it would be _especially_ tedious if you're trying to raise/lower land or water.

(A scenery scatter tool with maximum brush size would also nicely replace the old style random tree placement in the scenario editor, as mentioned in #16978 )

So I'd recommend increasing the maximum size of these brushes.

I _don't_ think the main brush size selection UI needs to be changed, though. It would be fine if the only way to get a brush bigger than 64x64 was to click on the center of the size selector and bring up the numeric entry dialog, which currently says "Please enter a size between 1 and 64" ... but would then say, "Please enter a size between 1 and 999"

Unless there's some technical reason that makes coding it extremely difficult, I don't see why we'd still be limiting maximum brush size to 64x64.

**Comment by IntelOrca** [comment]
Performance is the main issue. You will find it gets quite laggy if you increase the size, even more so for the mountain tool.

**Reply by ocalhoun6** [reply]
Depending on the situation, a bit of lag might be well worth it. Especially when it's hidden behind needing to manually enter the number so that unknowing users won't accidentally encounter this lag.

I'd definitely rather deal with lag than deal with clicking 244 times. It would be a _lot_ easier (and probably a lot faster) to raise the water on a 999x999 square once and wait for the lag than to raise a lagless 64x64 square two hundred times.

I guess if the lag is *too* extreme on the mountain tool, that one might stay restricted to a smaller size.

Maybe some trick could be done where the computer breaks the large square into smaller squares, then executes the action on the smaller squares one by one? I'm assuming that would take just as long, but at least the player would see things happening during that process, which would make it feel less laggy. Might actually look pretty cool to watch that happen. It would be kind of like doing it the slow way, 64 at a time ... but having the computer do it for you. (Though, again, that probably would be difficult or impossible to do with the mountain tool, so maybe that tool should be an exception.)

Also ... are _all_ of these tools laggy, or is it just the land/water? I guess depending on the number of objects involved, the scenery scatter or bulldozer tools might be much faster than the land/water tools in most cases, with the mountain tool lagging far, far behind in last place.

**Reply by spacek531** [reply]
If it takes more than 1 second to raise 999x999 one height then everyone will overshoot and take much longer

**Reply by IntelOrca** [reply]
When you move your cursor, it needs to calculate the cost of the action, this alone causes the lag. If you try and raise / lower the land, the lag may cause the user to move the land down or up far more than they wanted to.

**Reply by ocalhoun6** [reply]
Still ... I think I'd rather take my chances and carefully use the laggy tool than try to modify an entire 999x999 map with 64x64 brushes.

It could be tricky for the land/water tools, yes ... but as long as you move the mouse very carefully, it could save you a _lot_ of trouble vs. doing it the hard way. Maybe there could even be some way to disable user input while the calculation is underway? That would mean you could only raise/lower one unit at a time ... but it would prevent you from accidentally overshooting your goal.

(Or maybe there should be some warning message about how extremely large brush sizes may cause lag, so use with caution.)

But at least for the scenery scatter and bulldozer tools, the lag should be less of an issue -- since with those, you don't need to click and drag, you can just click once and let it do its thing.

**Reply by frutiemax** [reply]
I once wrote a plugin that restricted land permits for each player in a multiplayer server. I had to code a brush and it was really laggy when the brush was too large and the bigger the map was the bigger the lag it got.

**Reply by ocalhoun6** [reply]
Possibly part of this lag is because it's multiplayer and depending on network latency? Maybe it would be better in single player...

**Reply by Broxzier** [reply]
Selection is always singleplayer.
